hack/kserve_migration/kserve_migration.sh

```diff
@@ -0,0 +1,236 @@
+log INFO "removing owner references from virtual services"
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    vsvc_api_version=$(kubectl get virtualservices ${isvc_names[$i]} -n ${isvc_ns[$i]} -o json | jq --raw-output '.metadata.ownerReferences[0].apiVersion')
+    if [ "$vsvc_api_version" == "serving.kubeflow.org/v1beta1" ]; then
+        kubectl patch virtualservices ${isvc_names[$i]} -n ${isvc_ns[$i]} --type json -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
+    fi
+done
+sleep 5
+
+# Extract inference service uids
+log INFO "extracting inference service uids"
+declare -A infr_uid_map
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    infr_uid_map[${isvc_names[$i]}]=$(kubectl get inferenceservice.serving.kserve.io ${isvc_names[$i]} -n ${isvc_ns[$i]} -o json | jq --raw-output '.metadata.uid')
+done
+
+# Update knative services with new owner reference
+log INFO "updating knative services with new owner reference"
+for (( i=0; i<${ksvc_count}; i++ ));
+do
+    owner_ref_count=$(kubectl get ksvc ${ksvc_names[$i]} -n ${ksvc_ns[$i]} -o json | jq --raw-output '.metadata.ownerReferences | length')
+    if [ $owner_ref_count -eq 0 ]; then
+        isvc_name=${ksvc_isvc_map[${ksvc_names[$i]}]}
+        isvc_uid=${infr_uid_map[${isvc_name}]}
+        kubectl patch ksvc ${ksvc_names[$i]} -n ${ksvc_ns[$i]} --type='json' -p='[{"op": "add", "path": "/metadata/ownerReferences", "value": [{"apiVersion": "serving.kserve.io/v1beta1","blockOwnerDeletion": true,"controller": true,"kind": "InferenceService","name": "'${isvc_name}'","uid": "'${isvc_uid}'"}] }]'
+    fi
+done
+
+# Update virtual services with new owner reference
+log INFO "updating virtual services with new owner reference"
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    owner_ref_count=$(kubectl get virtualservices ${isvc_names[$i]} -n ${isvc_ns[$i]} -o json | jq --raw-output '.metadata.ownerReferences | length')
+    if [ $owner_ref_count -eq 0 ]; then
+        isvc_uid=${infr_uid_map[${isvc_names[$i]}]}
+        kubectl patch virtualservices ${isvc_names[$i]} -n ${isvc_ns[$i]} --type='json' -p='[{"op": "add", "path": "/metadata/ownerReferences", "value": [{"apiVersion": "serving.kserve.io/v1beta1","blockOwnerDeletion": true,"controller": true,"kind": "InferenceService","name": "'${isvc_names[$i]}'","uid": "'${isvc_uid}'"}] }]'
+    fi
+done
+sleep 5
+
+# Verify that all inference services are migrated and ready
+log INFO "verifying inference services are migrated and ready"
+for (( i=0; i<${isvc_count}; i++ ));
+do
+    (
+        trap 'log ERROR "inference service ${isvc_names[$i]} did not migrate properly. migration job exits with code 1."' ERR
+        kubectl wait --for=condition=ready --timeout=10s inferenceservice.serving.kserve.io/${isvc_names[$i]} -n ${isvc_ns[$i]}
+    )
+done
+
+# Start kfserving controller for clean up
+log INFO "starting kfserving controller for clean up"
+kubectl scale --replicas=1 statefulset.apps kfserving-controller-manager -n kfserving-system
```

Comment:
starting up the controller might have side effects, do we need to bring up the controller to clean up them?

Reply:
Yes @yuzisun . We need controller to clean up older inference service.

Reply:
would kfserving controller still reconciles the older inference service once it comes up ?

Reply:
yes @yuzisun

Reply:
So that can cause problem as two controllers are trying to update the knative service with race condition?

Reply:
@yuzisun I do not face any issues during testing. we are removing kfserving controller as well right after inference service clean up.

Reply:
ye I feel it might be safer to remove the finalizer so you do not need to scale back the controller to delete the resources